File: awkward.rb
module AwkwardMath

	def AwkwardMath.reload
		load 'awkward.rb'
	end

	class Point
		attr_reader :next_point

		def <<(vertex)
			set_next(vertex)
		end

		def set_next(vertex)
			@next_point = vertex
		end

		def inspect
			"Point: has next? #{!@next_point.nil?}"
		end

		def next
			@next_point
		end

	end

	class ActivePoint < Point 
		def inspect
			"ActivePoint: has next? #{!@next_point.nil?}"
		end
	end

	class Branch

		attr_reader :starting_point, :current_point

		def initialize(starting_point, current_point=nil)
			@starting_point = starting_point
			@current_point = current_point || starting_point
		end

		def <<(point)
			@current_point << point 
			@current_point = point
		end

		def dup
			starting_point = @starting_point.dup
			current = starting_point
			while ((next_point = current.next.dup) != nil)
				current << next_point
				current = next_point
			end
			Branch.new(starting_point, current)
		end

		def copy_and_close
			copy = self.dup
			copy << copy.starting_point
			Cycle.new(copy.current_point)
		end

		def length
			count = 1
			current = @starting_point
			while (current = current.next)
				count += 1
			end
			count
		end

	end

	class Cycle

		attr_reader :current_point

		def initialize(current_point)
			@starting_point = current_point
			@current_point = current_point
		end

		def active?
			@current_point.class == ActivePoint
		end

		def move
			@current_point = @current_point.next
		end

		def length
			count = 1
			current = @starting_point
			while (current = current.next) != @starting_point
				count += 1
			end
			count
		end

	end

	class StateMachine
		
		def next_state
			raise "Abstract method error"
		end

		def run_machine(iterations)
			iterations.times do 
				self.next_state
			end
			self
		end

	end

	class AwkwardStateMachine < StateMachine
		attr_reader :cycles, :branch

		# 
		# @param [int] intial_branch_length - The length of the initial
		# branch when disconnected from the activators
		def initialize(number_of_activators, initial_branch_length)
			
			activator = ActivePoint.new
			starting_point = activator
			(number_of_activators - 1).times do 
				next_activator = ActivePoint.new
				activator << next_activator
				activator = next_activator
			end

			current = activator
			initial_branch_length.times do 
				next_point = Point.new
				current << next_point
				current = next_point
			end

			@branch = Branch.new(starting_point, current)
			@cycles = []
		end

		def next_state
			# Advance the current cycles
			@cycles.each(&:move)
			unless self.active?
      	@cycles << @branch.copy_and_close 
      end
      @branch << Point.new
      self
		end

		def active?
			@cycles.any?(&:active?)
		end

		def lengths
			@cycles.map(&:length)
		end

	end

end

File: awkward.js
var Point = /** @class */ (function () {
    function Point(active) {
        if (active === void 0) { active = false; }
        this.active = active;
    }
    Point.prototype.setNext = function (vertex) {
        this.next = vertex;
    };
    Point.prototype.getNext = function () {
        if (!this.next)
            throw ("Doesn't have next");
        return this.next;
    };
    Point.prototype.hasNext = function () {
        return !!this.next;
    };
    Point.prototype.dup = function () {
        var p = new Point(this.active);
        if (this.hasNext())
            p.next = this.next;
        return p;
    };
    return Point;
}());
var Branch = /** @class */ (function () {
    function Branch(startingPoint, currentPoint) {
        if (currentPoint === void 0) { currentPoint = startingPoint; }
        this.startingPoint = startingPoint;
        this.currentPoint = currentPoint;
        if (!this.startingPoint)
            throw ("need starting point for Branch");
    }
    Branch.prototype.addPoint = function (point) {
        this.currentPoint.next = point;
        this.currentPoint = point;
    };
    Branch.prototype.dup = function () {
        var startingPoint = this.startingPoint.dup();
        var current = startingPoint;
        var nextPoint;
        while (current.hasNext()) {
            var next = current.next;
            nextPoint = next.dup();
            current.next = nextPoint;
            current = nextPoint;
        }
        return new Branch(startingPoint, current);
    };
    Branch.prototype.copyAndClose = function () {
        var copy = this.dup();
        copy.addPoint(copy.startingPoint);
        return new Cycle(copy.currentPoint);
    };
    Branch.prototype.length = function () {
        var count = 1;
        var current = this.startingPoint;
        while (current.hasNext()) {
            current = current.next;
            count++;
        }
        return count;
    };
    return Branch;
}());
var Cycle = /** @class */ (function () {
    function Cycle(currentPoint) {
        this.currentPoint = currentPoint;
        if (!(currentPoint instanceof Point))
            throw ('currentPoint must have a value');
        this.startingPoint = this.currentPoint;
    }
    Cycle.prototype.active = function () {
        return this.currentPoint.active;
    };
    Cycle.prototype.move = function () {
        this.currentPoint = this.currentPoint.next;
    };
    Cycle.prototype.length = function () {
        if (this._length)
            return this._length;
        var count = 1;
        var current = this.startingPoint;
        while (current.next !== this.startingPoint) {
            current = current.next;
            count++;
        }
        return this._length = count;
    };
    return Cycle;
}());
var Asm = /** @class */ (function () {
    function Asm(numActivators, numNonActivators) {
        this.numActivators = numActivators;
        this.numNonActivators = numNonActivators;
        this.cycles = [];
        this.branch = new Branch(new Point(true));
        for (var i = 1; i < this.numActivators; i++) {
            this.branch.addPoint(new Point(true));
        }
        for (var i = 0; i < this.numNonActivators; i++) {
            this.branch.addPoint(new Point());
        }
        this.cycles.push(this.branch.copyAndClose());
        this.branch.addPoint(new Point());
    }
    Asm.prototype.nextState = function () {
        this.cycles.forEach(function (c) { return c.move(); });
        if (!this.active()) {
            this.cycles.push(this.branch.copyAndClose());
        }
        this.branch.addPoint(new Point());
        return this;
    };
    Asm.prototype.runMachine = function (iterations) {
        for (var i = 1; i <= iterations; i++) {
            this.nextState();
        }
        return this;
    };
    Asm.prototype.active = function () {
        return this.cycles.some(function (c) { return c.active(); });
    };
    Asm.prototype.lengths = function () {
        return this.cycles.map(function (c) { return c.length(); });
    };
    return Asm;
}());
console.info('boot');
var a = new Asm(1, 1);
console.debug('made Asm', a);
console.debug('about to run machine');
a.runMachine(20);
console.debug('finished running machine');
console.info("Lengths: " + a.lengths());
